File: implementations/python-fastapi-pragmatic/src/presentation/api/v1/protected.py
from fastapi import APIRouter, Header, HTTPException

from src.presentation.schemas.common import ProtectedResponse

router = APIRouter(tags=["protected"])


@router.get("/protected", response_model=ProtectedResponse)
async def protected_endpoint(
    authorization: str | None = Header(default=None),
    x_request_id: str | None = Header(default=None),
):
    """Protected endpoint - middleware chain performance (auth, logging, validation)"""
    # Simulate authentication
    if authorization is None:
        raise HTTPException(status_code=401, detail="Authorization header required")

    if not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Invalid authorization format")

    token = authorization.replace("Bearer ", "")

    # Simulate token validation (in real app, would verify JWT)
    if len(token) < 10:
        raise HTTPException(status_code=401, detail="Invalid token")

    return ProtectedResponse(
        message="Access granted",
        user=f"user_from_token_{token[:8]}",
    )

File: implementations/typescript-express/src/routes/protected.ts
import { Router } from "express";

const router = Router();

router.get("/", (req, res) => {
  const authorization = req.headers.authorization;
  const xRequestId = req.headers["x-request-id"];

  // 1. Authorization 헤더 필수
  if (!authorization) {
    return res.status(401).json({ detail: "Authorization header required" });
  }

  // 2. Bearer 포맷 검증
  if (!authorization.startsWith("Bearer ")) {
    return res.status(401).json({ detail: "Invalid authorization format" });
  }

  const token = authorization.replace("Bearer ", "");

  // 3. 토큰 최소 길이 검증
  if (token.length < 10) {
    return res.status(401).json({ detail: "Invalid token" });
  }

  // 성공
  res.json({
    message: "Access granted",
    user: `user_from_token_${token.slice(0, 8)}`,
  });
});

export default router;
